### PR TITLE
new contract concept for distribution

### PR DESCRIPTION
**Pull request by geckoslair**

### CODE REVIEW

File: contracts/paratii/PTIDistributor.sol [hunk 1]
```diff
@@ -8,24 +8,45 @@ contract PTIDistributor is Ownable {
...
+    function() public payable { }
+
+    function distribute(address _toAddress, uint256 _amount, uint256 _salt, uint8 _v, bytes32 _r, bytes32 _s) {
+      bytes32 message = keccak256(_amount, _salt);
+      bytes memory prefix = "\x19Ethereum Signed Message:\n32";
+      bytes32 prefixedHash = keccak256(prefix, message);
+      require(!isUsed[_salt] && ecrecover(prefixedHash, _v, _r, _s) == owner);
+      isUsed[_salt] = true;
+      ParatiiToken token = ParatiiToken(registry.getContract('ParatiiToken'));
+      token.transfer(msg.sender, _amount);
```
Comment (ya7ya) [inline contracts/paratii/PTIDistributor.sol]: use `require` here to make sure the transfer happens or all changes are cancelled.
`require(token.transfer(msg.sender, _amount));`

File: contracts/paratii/PTIDistributor.sol [hunk 2]
```diff
@@ -8,24 +8,45 @@ contract PTIDistributor is Ownable {
...
         owner = msg.sender;
         registry = _registry;
     }
 
-    modifier onlyOwnerOrAvatar() {
-        require((msg.sender == owner) || (msg.sender == registry.getContract('Avatar')));
-        _;
+    function() public payable { }
+
+    function distribute(address _toAddress, uint256 _amount, uint256 _salt, uint8 _v, bytes32 _r, bytes32 _s) {
```
Comment (ya7ya) [inline contracts/paratii/PTIDistributor.sol]: what if we use a `bytes32` for the salt??

File: contracts/paratii/PTIDistributor.sol [hunk 3]
```diff
@@ -9,23 +9,43 @@ contract PTIDistributor is Ownable {
...
+    function() public payable { }
+
+    function distribute(address _toAddress, uint256 _amount, bytes32 _salt, string _reason, uint8 _v, bytes32 _r, bytes32 _s) {
+      bytes32 message = keccak256(_amount, _salt, _reason);
+      bytes memory prefix = "\x19Ethereum Signed Message:\n32";
+      bytes32 prefixedHash = keccak256(prefix, message);
+      require(!isUsed[_salt] && ecrecover(prefixedHash, _v, _r, _s) == owner);
+      isUsed[_salt] = true;
+      ParatiiToken token = ParatiiToken(registry.getContract('ParatiiToken'));
+      require(token.transfer(msg.sender, _amount));
```
Comment (jellegerbrandy) [inline contracts/paratii/PTIDistributor.sol]: you are tranfering the tokens to msg.sender but logging _toAddress.

File: contracts/paratii/PTIDistributor.sol [hunk 4]
```diff
@@ -9,23 +9,43 @@ contract PTIDistributor is Ownable {
...
+      LogDistribute(_toAddress, _amount, _reason);
     }
 
-    function() public payable { }
+    function checkOwnerPacked(uint256 _amount, bytes32 _salt, string _reason, uint8 _v, bytes32 _r, bytes32 _s) {
```
Comment (jellegerbrandy) [inline contracts/paratii/PTIDistributor.sol]: we need these checks?

File: contracts/paratii/PTIDistributor.sol (hunk 3 above)
Comment (geckoslair) [reply contracts/paratii/PTIDistributor.sol]: damn! you are right!

File: contracts/paratii/PTIDistributor.sol (hunk 4 above)
Comment (geckoslair) [reply contracts/paratii/PTIDistributor.sol]: don't know!

**Pull request merged**